Nodeset compiler/GDS

<!--
!ATTENTION!
Please read the following page carefully and provide us with all the
information requested:
https://github.com/open62541/open62541/wiki/Writing-Good-Issue-Reports

Use Github Markdown to format your text:
https://help.github.com/articles/basic-writing-and-formatting-syntax/

Fill out the sections and checklist below (add text at the end of each line).

!ATTENTION!
--------------------------------------------------------------------------------
-->

## Description

For GDS implementation, while trying to build the datatypes and nodeset from _open62541/deps/ua-nodeset/GDS/_, Nodeset Compiler reports an error.

## Background Information / Reproduction Steps

While addressing one of the review [comments](https://github.com/open62541/open62541/pull/2299#discussion_r245254082), this issue was found:

Used CMake options:

```bash
cmake ../ -DUA_NAMESPACE_ZERO="FULL" -DUA_BUILD_EXAMPLES=TRUE
```
Added below modification to examples/nodeset/CMakeLists.txt
```bash
# Generate types and namespace for GDS
ua_generate_nodeset_and_datatypes(
    NAME "gds"
    FILE_CSV "${FILE_NS_DIRPREFIX}/GDS/OpcUaGdsModel.csv"
    FILE_BSD "${FILE_NS_DIRPREFIX}/GDS/Opc.Ua.Gds.Types.bsd"
    NAMESPACE_IDX 4
    FILE_NS "${FILE_NS_DIRPREFIX}/GDS/Opc.Ua.Gds.NodeSet2.xml"
    INTERNAL
)
```
The error log:
```bash
Scanning dependencies of target open62541-generator-types-gds
[ 88%] Generating ../../src_generated/open62541/types_gds_generated.c, ../../src_generated/open62541/types_gds_generated.h, ../../src_generated/open62541/types_gds_generated_handling.h, ../../src_generated/open62541/types_gds_generated_encoding_binary.h
Traceback (most recent call last):
  File "/home/jayanth/gds_ns_generate/open62541/tools/generate_datatypes.py", line 516, in <module>
    parseTypeDefinitions(outname, f, args.namespace)
  File "/home/jayanth/gds_ns_generate/open62541/tools/generate_datatypes.py", line 358, in parseTypeDefinitions
    raise RuntimeError("Infinite loop detected trying to processing types " + name + ": unknonwn subtype " + str(unknownTypes(typeXml)))
RuntimeError: Infinite loop detected trying to processing types ApplicationRecordDataType: unknonwn subtype ['ApplicationType']
examples/nodeset/CMakeFiles/open62541-generator-types-gds.dir/build.make:65: recipe for target 'src_generated/open62541/types_gds_generated.c' failed
make[2]: *** [src_generated/open62541/types_gds_generated.c] Error 1
CMakeFiles/Makefile2:1365: recipe for target 'examples/nodeset/CMakeFiles/open62541-generator-types-gds.dir/all' failed
make[1]: *** [examples/nodeset/CMakeFiles/open62541-generator-types-gds.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

While building with python script
```bash
$ python generate_datatypes.py --namespace=2 --type-csv=../deps/ua-nodeset/GDS/OpcUaGdsModel.csv --type-bsd=../deps/ua-nodeset/GDS/Opc.Ua.Gds.Types.bsd --no-builtin ua_types_gds
```
The error log 
```bash
Traceback (most recent call last):
  File "generate_datatypes.py", line 516, in <module>
    parseTypeDefinitions(outname, f, args.namespace)
  File "generate_datatypes.py", line 358, in parseTypeDefinitions
    raise RuntimeError("Infinite loop detected trying to processing types " + name + ": unknonwn subtype " + str(unknownTypes(typeXml)))
RuntimeError: Infinite loop detected trying to processing types ApplicationRecordDataType: unknonwn subtype ['ApplicationType']
```
## Checklist

Please provide the following information:

 - [x] open62541 Version (release number or git tag): master
 - [ ] Other OPC UA SDKs used (client or server): 
 - [x] Operating system: Linux (Ubuntu 18.04)
 - [x] Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [ ] Wireshark network dump attached
 - [ ] Self-contained code example attached
 - [ ] Critical issue

--- a/tools/cmake/macros_public.cmake
+++ b/tools/cmake/macros_public.cmake
@@ -98,6 +98,10 @@ endfunction()
 #
 #   FILES_BSD        Path to the .bsd file containing the type definitions, e.g. 'Opc.Ua.Di.Types.bsd'. Multiple files can be
 #                   passed which will all combined to one resulting code.
+#   IMPORT_BSD      Combination of types array and path to the .bsd file containing additional type definitions referenced by
+#                   the FILES_BSD files. The value is separated with a hash sign, i.e.
+#                   'UA_TYPES#${PROJECT_SOURCE_DIR}/deps/ua-nodeset/Schema/Opc.Ua.Types.bsd'
+#                   Multiple files can be passed which will all be imported.
 #   [FILES_SELECTED] Optional path to a simple text file which contains a list of types which should be included in the generation.
 #                   The file should contain one type per line. Multiple files can be passed to this argument.
 #
@@ -105,7 +109,7 @@ endfunction()
 function(ua_generate_datatypes)
     set(options BUILTIN INTERNAL)
     set(oneValueArgs NAME TARGET_SUFFIX TARGET_PREFIX NAMESPACE_IDX OUTPUT_DIR FILE_CSV)
-    set(multiValueArgs FILES_BSD FILES_SELECTED)
+    set(multiValueArgs FILES_BSD IMPORT_BSD FILES_SELECTED)
     cmake_parse_arguments(UA_GEN_DT "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
 
     if(NOT DEFINED open62541_TOOLS_DIR)
@@ -160,6 +164,11 @@ function(ua_generate_datatypes)
         set(BSD_FILES_TMP ${BSD_FILES_TMP} "--type-bsd=${f}")
     endforeach()
 
+    set(IMPORT_BSD_TMP "")
+    foreach(f ${UA_GEN_DT_IMPORT_BSD})
+        set(IMPORT_BSD_TMP ${IMPORT_BSD_TMP} "--import=${f}")
+    endforeach()
+
     # Make sure that the output directory exists
     if(NOT EXISTS ${UA_GEN_DT_OUTPUT_DIR})
         file(MAKE_DIRECTORY ${UA_GEN_DT_OUTPUT_DIR})
@@ -177,6 +186,7 @@ function(ua_generate_datatypes)
         --namespace=${UA_GEN_DT_NAMESPACE_IDX}
         ${SELECTED_TYPES_TMP}
         ${BSD_FILES_TMP}
+        ${IMPORT_BSD_TMP}
         --type-csv=${UA_GEN_DT_FILE_CSV}
         ${UA_GEN_DT_NO_BUILTIN}
         ${UA_GEN_DT_INTERNAL_ARG}

--- a/tools/generate_datatypes.py
+++ b/tools/generate_datatypes.py
@@ -19,6 +19,7 @@ import json
 from nodeset_compiler.opaque_type_mapping import get_base_type_for_opaque as get_base_type_for_opaque_ns0
 
 types = OrderedDict() # contains types that were already parsed
+types_imported = OrderedDict() # contains types that were already parsed and marked as imported types (do not write to source code)
 typedescriptions = {} # contains type nodeids
 user_opaque_type_mapping = {} # contains user defined opaque type mapping
 
@@ -317,7 +318,7 @@ class StructType(Type):
 # Parse Typedefinitions #
 #########################
 
-def parseTypeDefinitions(outname, xmlDescription, namespace):
+def parseTypeDefinitions(outname, xmlDescription, namespace, addToTypes=None):
     def typeReady(element):
         "Are all member types defined?"
         for child in element:
@@ -364,15 +365,20 @@ def parseTypeDefinitions(outname, xmlDescription, namespace):
             if not typeReady(typeXml):
                 continue
             if name in builtin_types:
-                types[name] = BuiltinType(name)
+                newType = BuiltinType(name)
             elif typeXml.tag == "{http://opcfoundation.org/BinarySchema/}EnumeratedType":
-                types[name] = EnumerationType(outname, typeXml, namespace)
+                newType = EnumerationType(outname, typeXml, namespace)
             elif typeXml.tag == "{http://opcfoundation.org/BinarySchema/}OpaqueType":
-                types[name] = OpaqueType(outname, typeXml, namespace, get_base_type_for_opaque(name)['name'])
+                newType = OpaqueType(outname, typeXml, namespace, get_base_type_for_opaque(name)['name'])
             elif typeXml.tag == "{http://opcfoundation.org/BinarySchema/}StructuredType":
-                types[name] = StructType(outname, typeXml, namespace)
+                newType = StructType(outname, typeXml, namespace)
             else:
                 raise Exception("Type not known")
+
+            types[name] = newType
+            if addToTypes is not None:
+                addToTypes[name] = newType
+
             del snippets[name]
 
 ##########################
@@ -492,6 +498,14 @@ parser.add_argument('-t', '--type-bsd',
                     default=[],
                     help='bsd file with type definitions')
 
+parser.add_argument('-i', '--import',
+                    metavar="<importBsds>",
+                    type=str,
+                    dest="import_bsd",
+                    action='append',
+                    default=[],
+                    help='combination of TYPE_ARRAY#filepath.bsd with type definitions which should be loaded but not exported/generated')
+
 parser.add_argument('outfile',
                     metavar='<outputFile>',
                     help='output file w/o extension')
@@ -511,6 +525,13 @@ for builtin in builtin_types:
 
 for f in args.opaque_map:
     user_opaque_type_mapping.update(json.load(f))
+
+for i in args.import_bsd:
+    (outname_import, file_import) = i.split("#")
+    outname_import = outname_import.lower()
+    if outname_import.startswith("ua_"):
+        outname_import = outname_import[3:]
+    parseTypeDefinitions(outname_import, file_import, args.namespace, addToTypes=types_imported)
 
 for f in args.type_bsd:
     parseTypeDefinitions(outname, f, args.namespace)
@@ -554,6 +575,7 @@ def iter_types(v):
         l = list(filter(lambda t: t.name in selected_types, l))
     if args.no_builtin:
         l = list(filter(lambda t: type(t) != BuiltinType, l))
+    l = list(filter(lambda t: t.name not in types_imported, l))
     return l
 
 ################

